### PR TITLE
Enhance security of scaffolded apps

### PR DESCRIPTION
**Pull request by gunjpan**
WIP: for **Apply "Best Practices for Express in Production" in project template**: https://github.com/strongloop-internal/scrum-loopback/issues/583 

Connect to strongloop-internal/scrum-loopback#583

### CODE REVIEW

File: templates/components/api-server/component.js [hunk 1]
```diff
@@ -14,6 +14,7 @@ template.package = {
   "dependencies": {
     "compression": "^1.0.3",
     "cors": "^2.5.2",
+    'helmet': "^0.14.0",
```
Comment (bajtos) [inline templates/components/api-server/component.js]: Please use double-quotes here for consistency.

File: templates/components/api-server/template/server/server.js [hunk 2]
```diff
@@ -1,7 +1,11 @@
 var loopback = require('loopback');
 var boot = require('loopback-boot');
+var helmet = require('helmet');
 
 var app = module.exports = loopback();
+//use helmet to improve app security: more details on Best practices for Express in Production:
+// https://strongloop.com/strongblog/best-practices-for-express-in-production-part-one-security/
+app.use(helmet());
```
Comment (bajtos) [inline templates/components/api-server/template/server/server.js]: All middleware should be configured via `server/middleware.json`. To make it easier for users to tweak helmet config, the scaffolded `server/middleware.json` should include configuration mirroring helmet's defaults. That way users can just change a flag value, instead of having to look up all flag names in helmet docs.

**Comment by gunjpan** [comment]
@bajtos : Please review. Moved Helmet config to middleware.json template and excluded CSP (test cases to follow soon). 
-  I've set some default(i.e. random) values for hpkp. We can instruct loopback-generator to modify hpkp settings. Your take on those hpkp default values, please. Thank you.

File: templates/components/api-server/template/server/middleware.json [hunk 3]
```diff
@@ -10,6 +10,26 @@
         "credentials": true,
         "maxAge": 86400
       }
+    },
+    "helmet#xssFilter": {
+      "params": {
+        "setOnOldIE": true
```
Comment (bajtos) [inline templates/components/api-server/template/server/middleware.json]: > On old versions of Internet Explorer, this creates a vulnerability (see here and here), and so the header is set to 0 to disable it. To force the header on all versions of IE, add the option:

I believe we should not enable this option by default, as it would create a security vulnerability.

File: templates/components/api-server/template/server/middleware.json [hunk 4]
```diff
@@ -10,6 +10,26 @@
...
+      }
+    },
+    "helmet#frameguard": {
+      "params": "deny"
```
Comment (bajtos) [inline templates/components/api-server/template/server/middleware.json]: I'd prefer to wrap the value in an array, to make it more clear how to pass more than one item.

```js
// default in the scaffolding
"params": ["deny"]
// an example of a custom config
"params": ["allow-from", "http://example.com"]
```

File: templates/components/api-server/template/server/middleware.json [hunk 5]
```diff
@@ -10,6 +10,26 @@
...
+    },
+    "helmet#hsts": {
+      "params": {
+        "maxAge": 0,
+        "includeSubdomains": true
+      }
+    },
```
Comment (bajtos) [inline templates/components/api-server/template/server/middleware.json]: Ha, I didn't realise "hsts" sends the header only for HTTPs requests. That's great, as it allows us to include the middleware by default :+1:

**Comment by bajtos** [comment]
@gunjpan thank you for the pull request, the proposed changes look mostly good. Could you please rebase this patch on top of the current master?

Also could you please add automated tests to verify the implementation? I think they should go to [test/end-to-end.js](https://github.com/strongloop/loopback-workspace/blob/07b1cec1d7b773cf967a1c91c0022af6dfc80121/test/end-to-end.js) and simply verify that the app serves expected security headers.

File: test/end-to-end.js [hunk 6]
```diff
@@ -97,6 +98,20 @@ describe('end-to-end', function() {
...
+       .expect('x-xss-protection', '1; mode=block')
+       .expect('x-content-type-options', 'nosniff')
+       .expect('x-download-options', 'noopen')
+       .expect(function(res){
+          var headers = res.headers;
+          console.log(headers);
+          headers.should.not.have.property('x-powered-by');
+       })
+       .expect(200, done);
+    });
```
Comment (gunjpan) [inline test/end-to-end.js]: @bajtos : Please review

File: test/end-to-end.js [hunk 7]
```diff
@@ -97,6 +98,20 @@ describe('end-to-end', function() {
...
+    
+    it('provides security headers for all URLs ', function(done){
+      request(app).get('/')
+       .expect('X-frame-options', 'DENY')
+       .expect('x-xss-protection', '1; mode=block')
+       .expect('x-content-type-options', 'nosniff')
+       .expect('x-download-options', 'noopen')
+       .expect(function(res){
+          var headers = res.headers;
+          console.log(headers);
```
Comment (bajtos) [inline test/end-to-end.js]: Is this log here to work around insufficient error details printed by `headers.should.not.have.property`?

I am proposing to use `debug` instead.

```js
// at the top
var debug = require('debug')('test');

// here:
debug(headers);
```

File: test/end-to-end.js [hunk 8]
```diff
@@ -731,7 +746,7 @@ function execNpm(args, options, cb) {
 
 function installSandboxPackages(cb) {
   this.timeout(300 * 1000);
-  install(SANDBOX, PKG_CACHE, ['dependencies', 'devDependencies'], cb);
+    install(SANDBOX, PKG_CACHE, ['dependencies', 'devDependencies'], cb);
```
Comment (bajtos) [inline test/end-to-end.js]: This looks like an unrelated whitespace change to me, please revert.

**Comment by bajtos** [comment]
One minor comment, the patch LGTM otherwise.

File: test/end-to-end.js (hunk 7 above)
Comment (gunjpan) [inline test/end-to-end.js]: @bajtos : Grrrr... I forgot to remove `console.log`, it was for my own reference only...
I believe, `headers.should.not.have.property` throws enough error details, I will remove the log part.

**Comment by bajtos** [comment]
Landed, thank you!
